libc/intrin/interrupts.S

```diff
@@ -0,0 +1,173 @@
+│ of this software dedicate any and all copyright interest in the              │
+│ software to the public domain. We make this dedication for the benefit       │
+│ of the public at large and to the detriment of our heirs and                 │
+│ successors. We intend this dedication to be an overt act of                  │
+│ relinquishment in perpetuity of all present and future rights to this        │
+│ software under copyright law.                                                │
+│                                                                              │
+│ THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,              │
+│ EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF           │
+│ MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.       │
+│ IN NO EVENT SHALL THE AUTHORS BE LIABLE FOR ANY CLAIM, DAMAGES OR            │
+│ OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,        │
+│ ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR        │
+│ OTHER DEALINGS IN THE SOFTWARE.                                              │
+╚─────────────────────────────────────────────────────────────────────────────*/
+#include "libc/dce.h"
+#include "libc/macros.internal.h"
+#include "libc/runtime/pc.internal.h"
+
+//	Code and data structures for bare metal interrupt handling.
+
+#define ISR_STK_SZ 0x10000
+#define ISR_STK_ALIGN 0x10
+//	Interrupt stack to use for IRQs.  TODO.
+#define IRQ_IST   1
+//	Interrupt stack to use for CPU exceptions.
+#define EXCEP_IST 2
+
+//	Interrupt numbers to use for IRQs 0 & 8.  TODO: implement these!
+#define IRQ0 0x20
+#define IRQ8 0x28
+
+	.init.start 100,_init_isr
+	push	%rdi
+	call	isr_init
+	pop	%rdi
+	.init.end 100,_init_isr
+
+//	Interrupt service routines for CPU exceptions 0—31.
+	i = 31
+	.rept	30
+	push	%rsi			# preserve rsi
+	mov	$i,%sil			# rsi = exception number
+1:	jmp	1f			# 🦘
+	i = i - 1
+	.endr
+__excep1_isr:
+	push	%rsi
+	mov	$1,%sil
+1:	jmp	1f
+__excep0_isr:
+	push	%rsi
+	xor	%esi,%esi
+1:	test	$8,%esp			# if no error code was pushed,
+	jnz	2f			# stuff our own
+	pushq	(%rsp)
+	orq	$-1,8(%rsp)
+2:	movzbq	%sil,%rsi		# zero-extend the exception number
+	push	%rcx			# preserve registers which we will
+	push	%rdx			# use to call kprintf
+	push	%r8
+	mov	40(%rsp),%rcx		# edx:rcx = "caller" cs:rip
+	mov	48(%rsp),%edx
+	mov	32(%rsp),%r8		# r8 = error code
+	push	%rax			# preserve other call-used registers
+	push	%rdi
+	push	%r9
+	push	%r10
+	push	%r11
+	mov	%ss,%eax		# preserve ds, es, ss
+	push	%rax
+	mov	%ds,%eax
+	push	%rax
+	mov	%es,%eax
+	push	%rax
+	mov	$GDT_LONG_DATA,%eax	# ...& load ds, es, ss correctly
+	mov	%eax,%ss
+	mov	%eax,%ds
+	mov	%eax,%es
+	ezlea	.excep_msg,di		# stack should be 16-byte aligned now
+	xor	%eax,%eax		# kprintf is variadic; remember to
+					# pass no. of vector regs. used (= 0)
+	call	kprintf			# print error message
```

Comment:
Is this going to force a dependency on kprintf() into o/tiny/examples/life.com? If so, we might want to do a one-off error report.

Reply:
Hello @jart,

> Is this going to force a dependency on kprintf() into o/tiny/examples/life.com?

Not yet.  I am trying to arrange to link in the IDT logic only when the program is already linking in some sort of crash debugging (I am still not familiar with how Cosmopolitan's crash debugging capability works).

Thank you!